[SMP-2400] WIP Harness Self-Managed EE support bundle utility (#5011)
harness/developer-hub: docs/self-managed-enterprise-edition/support-bundle-utility.md

        selector:
          - <YOUR_APP_SELECTOR_LABEL>
```

:::info note
The tool by default redacts any password or secrets with common regex patterns.

:::

### Download the example manifest

An example support bundle manifest is available in the [Harness Helm chart repo](https://github.com/harness/helm-charts/tree/main/support-bundle-manifests#readme).

After you have [installed the utility](#installation-options), download and prepare your support bundle manifest.

To download the support bundle manifest and prepare it for use, do the following:

1. Run the following command.

   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/script.sh) <YOUR_NAMESPAACE> <YOUR_RELEASE_NAME>
   ```

   This will create a `support-bundle.yaml` file in the current directory. You can use this file to collect the support bundle information from all of your running services.

2. To download a module-specific manifest, you can use the following command.

   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/script.sh) <your-namespace> <your-release-name> <module-name>
   ```

> 

Author: ```suggestion
### Services

The `support-bundle.yaml` file includes the following services.

- access-control
- change-data-capture
- debezium-service
- gateway
- manager
- log-service
- migrator
- ng-custom-dashboards
- ng-dashboard-aggregator
- ng-manager
- pipeline-service
- platform-service
- policy-mgmt
- scm
- service-discovery-manager
- template-service

```

Reviewer: It has all the services, so this will not be accurate. I guess we can skip this comment (https://github.com/harness/developer-hub/pull/5011#discussion_r1466697189), the customer need not know which services supports what kind of logging. Everything else looks good

- For services with file logging enabled, Harness fetches the log files. You can use the following script to configure the number of log files.

   ```yaml
   apiVersion: troubleshoot.sh/v1beta2
   kind: SupportBundle
   metadata:
     name: supportbundle
   spec:
     collectors:
       - exec:
           name: logs/<SERVICE_NAME>
           selector:
             - app.kubernetes.io/name=<SERVICE_NAME>
           namespace: <your-namespace>
           command:
             - "/bin/bash"
             - "-c"
             - |
               logs_path="/opt/harness/logs"
               num_files=2 # <--- define number of files required

               for file in $(ls $logs_path/pod*.log | sort -n -k1,1 | tail -n $num_files); do
                   echo "==> $file &lt;=="
                   cat "$file"
                done
            timeout: 100s
   ```

- For services without file logging, Harness collects Kubernetes logs. 

   ```yaml
    apiVersion: troubleshoot.sh/v1beta2
    kind: SupportBundle
    metadata:
      name: supportbundle
    spec:
      collectors:
       - logs:
           selector:
             - app.kubernetes.io/name=<SERVICE_NAME>
   ```

> :::info

Author: ```suggestion
```

> Services without file logging have a maximum of 10MB logs available.

Author: ```suggestion
```

> 

Author: ```suggestion
```

> :::

Author: ```suggestion
```

      namespace: <namespace> # <-- Optional, if not provided it will search in all namespaces
      release-name: <name of the release> # <-- Optional. Note: If this is provided, namespace is mandatory
```

### Collect ConfigMap data

To collect ConfigMap data, use the following configuration.

```yaml
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  name: supportbundle
spec:
  collectors:
    - configMap:
        selector:
          - <YOUR_APP_SELECTOR_LABEL>
```

:::info note
The tool by default redacts any password or secrets with common regex patterns.

:::

### Download the example manifest

An example support bundle manifest is available in the [Harness Helm chart repo](https://github.com/harness/helm-charts/tree/main/support-bundle-manifests#readme).

After you have [installed the utility](#installation-options), download and prepare your support bundle manifest.

To download the support bundle manifest and prepare it for use, do the following:

1. Run the following command.

   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/script.sh) <YOUR_NAMESPAACE> <YOUR_RELEASE_NAME>
   ```

>    This will create a `support-bundle.yaml` file in the current directory. You can use this file to collect the support bundle.

Reviewer: Can we add that this will create a support bundle that will collect information from all the running services? (and not just limited to a single module). We will want customers to use this approach mostly

>    This will create a `support-bundle.yaml` file in the current directory. You can use this file to collect the support bundle.

Author: ```suggestion
   This will create a `support-bundle.yaml` file in the current directory. You can use this file to collect the support bundle information from all of your running services.
```

    - configMap:
        selector:
          - <YOUR_APP_SELECTOR_LABEL>
```

:::info note
The tool by default redacts any password or secrets with common regex patterns.

:::

### Download the example manifest

An example support bundle manifest is available in the [Harness Helm chart repo](https://github.com/harness/helm-charts/tree/main/support-bundle-manifests#readme).

After you have [installed the utility](#installation-options), download and prepare your support bundle manifest.

#### Pre-requisite

`yq` v4 or above. For installation instructions, go to [Install](https://github.com/mikefarah/yq?tab=readme-ov-file#install).

To download the support bundle manifest and prepare it for use, do the following:

- Run the following command applicable for your OS.
   
   **Linux**
   
   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/scripts/linux.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME>
   ```

   **MacOS**
   
   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/scripts/macos.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME>
   ```

   **Windows**
   

>    ```bash
>    bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/scripts/windows.ps1) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME>
>    ```

Reviewer: ```suggestion
Support to be added
```

I am testing the script for windows, let's add it after it is tested.

   This will create a `support-bundle.yaml` file in the current directory. You can use this file to collect the support bundle information from all of your running services.

### Download a module-specific manifest

- To download a module-specific manifest, you can use the following command.

   ```bash

>    bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/script.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME> <MODULE_NAME>

Reviewer: ```suggestion
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/script/linux.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME> <MODULE_NAME>
```

   ```

   **Time Since**

   You can provide time since for the logs to be collected.

   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/scripts/linux.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME> --last 1 hours
   ```

   Supported values are:
   
   - x minutes
   - x hours
   - x days

   **Time Range**

   You can provide a time range for the logs to be collected. Provide the start and end date in YYYY-MM-DD format.

   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/scripts/linux.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME> --between 2021-01-01 2021-01-02
   ```

   **Number of Files**

>    You can provide the number of files to be collected for the logs. The default value is 2.

Reviewer: ```suggestion
   You can provide the number of files to be collected for the logs. The default value is 2. This is exclusive of `--last` and `--between` flags, if either is provided this will not be used.
```

   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/scripts/linux.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME> --num-files 10
   ```

   **File Path**
   

>    If you are using a different file path for the logs, you can provide the path using the following command.

Reviewer: ```suggestion
   If you are using a different file path for the logs, you can provide the path using the following command. Adding `*.log` or any extension you have provided for the file is important. If you have provided the file path as /logs/service.txt then provide the path in the flag as /logs/service*.txt
```

   ```bash
   bash <(curl -sSL https://raw.githubusercontent.com/harness/helm-charts/main/support-bundle-manifests/scripts/linux.sh) <YOUR_NAMESPACE> <YOUR_RELEASE_NAME> --filepath /path/to/logs*.log
   ```

#### Example manifest

> The following example manifest collects Helm values for releases and logs ConfigMap data for the `ng-manager` service.

Reviewer: ```suggestion
The following example manifest collects Helm values for releases, logs and ConfigMap data for the `ng-manager` service.
```

```yaml
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  name: supportbundle
spec:
  collectors:
    - clusterResources:
        exclude: true
    - exec:
        name: logs/platform/ng-manager
        selector:
          - app.kubernetes.io/name=ng-manager
        namespace: YOUR_NAMESPACE
        command:
          - "/bin/bash"
          - "-c"
          - |
            logs_path="/opt/harness/logs"
            num_files=2

            for file in $(ls $logs_path/pod*.log | sort -n -k1,1 | tail -n $num_files); do
                echo "==> $file &lt;=="
                cat "$file"
            done
        timeout: 100s
```

## Generate the support bundle

> The process of generating the support bundle involves collecting diagnostic information as required by the user. This process involves interacting with various system components to gather logs, ConfigMaps (excluding secrets), and override files. The support bundle generator is optimized to handle concurrent requests from multiple users and ensure efficient use of resources.

Reviewer: ```suggestion
The process of generating the support bundle involves collecting diagnostic information as required by the user. This process involves interacting with various system components to gather logs, ConfigMaps (excluding secrets), and override files. The support bundle generator is optimized to use minimum resources to collect all the required data without having any effect on the running services.
```
Concurrent requests/users is invaluable here, we wouldn't expect multiple users on the same system to run this utility.

   ```

- If you have installed the tool by downloading the binary, then run the following command to generate the support bundle.

   ```
   ./collect /path/to/support-bundle.yaml
   ```

This command will generate a compressed file named `support-bundle-*.tar.gz`. Extract the file to access collected data.

### Location of collected data

The data collected using the provided example manifest will be stored in the directories listed below.

- Helm release data is included the `helm` folder.

- Cluster resources data is inside the `cluster-resource` folder. This includes the deployment, sts, and pod data.

- Logs are inside the `logs` folder.

## Send logs to Harness

To send your log bundle files to Harness, do the following:

1. Go to [Harness secure file drop](https://harness.sendsafely.com/d/support).
2. Enter your email address, and then select **Submit**.
3. Complete the MFA.
4. Drag your file(s) or click to add the file manually for upload.
5. Select **Submit**.

## Use in an air gap environment

To use the support bundle utility in an air gap environment, do the following:

1. Download the tool binary.
2. Download the manifest.

> 3. Transfer the manifest to the jumpbox VM if the VM doesn't have internet access. If the VM has internet access, you can run the command directly run the command directly.

Reviewer: ```suggestion
3. Transfer the manifest to the jumpbox VM if the VM doesn't have internet access. If the VM has internet access, you can run the above commands directly on the VM.
```
